bip-musig2.mediawiki

```diff
@@ -410,6 +391,49 @@ The reference implementation is for demonstration purposes only and not to be us
 
 == Remarks on Security and Correctness ==
 
+=== Modifications to Nonce Generation ===
+
+Implementors must avoid modifying the ''NonceGen'' algorithm without being fully aware of the implications.
+The following two modifications are secure when applied correctly and may be useful in special circumstances, e.g., in very restricted environments where secure randomness is not available.
+
+First, on systems where obtaining uniformly random values is much harder than maintaining a global atomic counter, it can be beneficial to modify ''NonceGen''.
+Instead of drawing ''rand' '' uniformly at random, ''rand' '' can be the value of an atomic counter that is incremented whenever it is read.
+With this modification, the secret signing key ''sk'' of the signer generating the nonce is '''not''' an optional argument and must be provided to ''NonceGen''.
+The security of the resulting scheme is then depending on the requirement that the counter must never return the same output in two ''NonceGen'' invocations with the same ''sk''.
+
+Second, if there is a unique signer who is supposed to send the ''pubnonce'' last, it is possible to modify nonce generation for this single signer to not require high-quality randomness.
+Such a nonce generation algorithm ''DeterministicSign'' is specified below.
+Note that there are no optional arguments except ''rand'', which can be ommitted if randomness is entirely unavailable.
+''DeterministicSign'' requires the argument ''othernonce'' which is the output of ''NonceAgg'' run on the ''pubnonce'' value of all other signers.
+Hence, using ''DeterministicSign'' is only possible for the last signer to generate a nonce and makes the signer stateless, similar to the stateless signer described in the [[#nonce-generation|Nonce Generation]] section.
+
+==== Deterministic Signing ====
+
+Input:
+* The secret signing key ''sk'': a 32-byte array
+* The aggregate public nonce ''othernonce'': a 66-byte array
+* The number ''u'' of public keys with ''0 < u < 2^32''
+* The public keys ''pk<sub>1..u</sub>'': ''u'' 32-byte arrays
+* The number ''v'' of tweaks with ''0 &le; v < 2^32''
+* The tweaks ''tweak<sub>1..v</sub>'': ''v'' 32-byte arrays
+* The tweak methods ''is_xonly_t<sub>1..v</sub>'' : ''v'' booleans
+* The message ''m'': a 32-byte array
+* The auxiliary randomness ''rand'': a 32-byte array or 0-byte array (optional argument)
+
+'''''DeterministicSign(sk, othernonce, pk<sub>1..u</sub>, tweak<sub>1..v</sub>, is_xonly_t<sub>1..v</sub>, m, rand)''''':
+* If ''len(rand) > 0:''
+** Let ''sk' '' be the byte-wise xor of ''sk'' and ''hash<sub>MuSig/aux</sub>(rand)''
+* Else: let ''sk' = sk''
+* Let aggpk = ''KeyAgg(pk<sub>1..u</sub>, tweak<sub>1..v</sub>, is_xonly_t<sub>1..v</sub>)''
+* Let ''k<sub>i</sub> = int(hash<sub>MuSig/stateless/nonce</sub>(sk' || othernonce || aggpk || m || bytes(1, i))) mod n'' for ''i = 1,2''
```

Comment:
```suggestion
* Let ''k<sub>i</sub> = int(hash<sub>MuSig/deterministic/nonce</sub>(sk' || othernonce || aggpk || m || bytes(1, i))) mod n'' for ''i = 1,2''
```

I think if we call the algorithm DeterministicSign (which is a good name), that's slightly better. (It won't matter much though if we unify the two methods of stateless signing.)